// @silkey/sdk/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.tokenPayloadVerifier = exports.messageToSign = exports.generateSSORequestParams = exports.fetchSilkeyEthAddress = exports.settings = exports.Verifier = exports.blockchain = exports.utils = exports.Models = void 0;
exports.Models = __importStar(require("./models"));
exports.utils = __importStar(require("./utils/helpers"));
exports.blockchain = __importStar(require("./contracts"));
var Verifier_1 = require("./services/Verifier");
Object.defineProperty(exports, "Verifier", { enumerable: true, get: function () { return Verifier_1.Verifier; } });
var settings_1 = require("./config/settings"); // do not change path!
Object.defineProperty(exports, "settings", { enumerable: true, get: function () { return settings_1.settings; } });
var sso_1 = require("./services/sso");
Object.defineProperty(exports, "fetchSilkeyEthAddress", { enumerable: true, get: function () { return sso_1.fetchSilkeyEthAddress; } });
Object.defineProperty(exports, "generateSSORequestParams", { enumerable: true, get: function () { return sso_1.generateSSORequestParams; } });
Object.defineProperty(exports, "messageToSign", { enumerable: true, get: function () { return sso_1.messageToSign; } });
Object.defineProperty(exports, "tokenPayloadVerifier", { enumerable: true, get: function () { return sso_1.tokenPayloadVerifier; } });
